libbz2-rs-sys/src/blocksort.rs

```diff
@@ -390,209 +390,57 @@ fn mainGtU(
-        s2 = quadrant[i2 as usize];
-        if s1 != s2 {
-            return s1 > s2;
-        }
-        i1 = i1.wrapping_add(1);
-        i2 = i2.wrapping_add(1);
-        c1 = block[i1 as usize];
-        c2 = block[i2 as usize];
-        if c1 != c2 {
-            return c1 > c2;
-        }
-        s1 = quadrant[i1 as usize];
-        s2 = quadrant[i2 as usize];
-        if s1 != s2 {
-            return s1 > s2;
-        }
-        i1 = i1.wrapping_add(1);
-        i2 = i2.wrapping_add(1);
-        c1 = block[i1 as usize];
-        c2 = block[i2 as usize];
-        if c1 != c2 {
-            return c1 > c2;
-        }
-        s1 = quadrant[i1 as usize];
-        s2 = quadrant[i2 as usize];
-        if s1 != s2 {
-            return s1 > s2;
-        }
-        i1 = i1.wrapping_add(1);
-        i2 = i2.wrapping_add(1);
-        c1 = block[i1 as usize];
-        c2 = block[i2 as usize];
-        if c1 != c2 {
-            return c1 > c2;
-        }
-        s1 = quadrant[i1 as usize];
-        s2 = quadrant[i2 as usize];
-        if s1 != s2 {
-            return s1 > s2;
-        }
-        i1 = i1.wrapping_add(1);
-        i2 = i2.wrapping_add(1);
-        c1 = block[i1 as usize];
-        c2 = block[i2 as usize];
-        if c1 != c2 {
-            return c1 > c2;
-        }
-        s1 = quadrant[i1 as usize];
-        s2 = quadrant[i2 as usize];
-        if s1 != s2 {
-            return s1 > s2;
-        }
-        i1 = i1.wrapping_add(1);
-        i2 = i2.wrapping_add(1);
-        c1 = block[i1 as usize];
-        c2 = block[i2 as usize];
-        if c1 != c2 {
-            return c1 > c2;
-        }
-        s1 = quadrant[i1 as usize];
-        s2 = quadrant[i2 as usize];
-        if s1 != s2 {
-            return s1 > s2;
-        }
-        i1 = i1.wrapping_add(1);
-        i2 = i2.wrapping_add(1);
-        c1 = block[i1 as usize];
-        c2 = block[i2 as usize];
+    let chunk1 = &block[i1 as usize..][..12];
+    let chunk2 = &block[i2 as usize..][..12];
+
+    for (c1, c2) in chunk1.chunks_exact(4).zip(chunk2.chunks_exact(4)) {
+        let c1 = u32::from_be_bytes(c1[..4].try_into().unwrap());
+        let c2 = u32::from_be_bytes(c2[..4].try_into().unwrap());
+
         if c1 != c2 {
             return c1 > c2;
         }
-        s1 = quadrant[i1 as usize];
-        s2 = quadrant[i2 as usize];
-        if s1 != s2 {
-            return s1 > s2;
+    }
+
+    i1 += 12;
+    i2 += 12;
+
+    for _ in 0..nblock.div_ceil(8) {
+        let b1 = &block[i1 as usize..][..8];
+        let b2 = &block[i2 as usize..][..8];
+
+        let q1 = &quadrant[i1 as usize..][..8];
+        let q2 = &quadrant[i2 as usize..][..8];
+
+        if b1 != b2 || q1 != q2 {
+            for (((c1, c2), s1), s2) in b1.iter().zip(b2).zip(q1).zip(q2) {
+                if c1 != c2 {
+                    return c1 > c2;
+                }
+                if s1 != s2 {
+                    return s1 > s2;
+                }
+            }
```

Comment:
I haven't yet found a good way to vectorize this part (tried some stuff with xor and `leading_zeros` but I could not get it to be correct so far). at least it's out of the hot path, and the equality check will uses full-width loads/compares (even an avx one for the quadrant compare)